aptos-move/framework/src/natives/cryptography/bls12381.rs

```diff
@@ -0,0 +1,696 @@
+        None => {
+            return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)]));
+        }
+    };
+
+    // NOTE(Gas): 2 bilinear pairings and a hash-to-curve
+    let verify_result = signature_verify_with_gas(&sig, &pk, msg_bytes, &mut cost, gas_params);
+
+    Ok(NativeResult::ok(
+        cost,
+        smallvec![Value::bool(verify_result)],
+    ))
+}
+
+/***************************************************************************************************
+ * native fun bls12381_aggregate_pop_verified_pubkeys
+ *
+ *   gas cost: base_cost + per_pubkey_deserialize_cost * min(num_validatable_pubkeys + 1, num_pubkeys)
+ *                       +? per_pubkey_aggregate_cost * num_pubkeys
+ *
+ * where +? indicates that the expression stops evaluating there if the previous gas-charging step
+ * failed, num_pubkeys is the # of public keys given as input, and num_validatable_pubkeys is the
+ * # of public keys that deserialize successfully.
+ *
+ * NOTE(ValidatablePK): We refer to the public keys that deserialize correctly as "validatable"
+ * above, since successful deserialization is not a sufficient criteria for the "validatability"
+ * of a PK: e.g., the PK could still be susceptible to small-subgroup attacks or rogue-key attacks.
+ *
+ * NOTE: If all PKs deserialize, then per_pubkey_deserialize_cost is charged num_pubkeys times.
+ * Otherwise, if only num_validatable_pubkeys deserialize correctly, an extra per_pubkey_deserialize_cost
+ * must be charged for the failed deserialization.
+ **************************************************************************************************/
+fn native_bls12381_aggregate_pop_verified_pubkeys(
+    gas_params: &GasParameters,
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    // Parses a Vec<Vec<u8>> of all serialized public keys
+    let pks_bytes = pop_vec_of_vec_u8(&mut arguments)?;
+    let num_pks = pks_bytes.len();
+    let mut cost = gas_params.base_cost;
+
+    // If zero PKs were given as input, return None.
+    if pks_bytes.is_empty() {
+        return Ok(NativeResult::ok(
+            cost,
+            smallvec![Value::vector_u8(vec![]), Value::bool(false)],
+        ));
+    }
+
+    let pks = bls12381_deserialize_pks_with_gas(pks_bytes, &mut cost, gas_params);
+    debug_assert!(pks.len() <= num_pks);
+
+    // If not all PKs were successfully deserialized, return None and only charge for the actual work done
+    if pks.len() != num_pks {
+        return Ok(NativeResult::ok(
+            cost,
+            smallvec![Value::vector_u8(vec![]), Value::bool(false)],
+        ));
+    }
+
+    // Aggregate the public keys (this will NOT subgroup-check the individual PKs)
+    // NOTE(Gas): |pks| elliptic curve additions
+    cost += gas_params.per_pubkey_aggregate_cost * num_pks as u64;
```

Comment:
any reason this is directly in the native function? because it only appears once?

Reply:
Yes, it appears only once. Is there a better pattern here?